### PR TITLE
Add python version of tilers

### PR DESCRIPTION
**Pull request by sovrasov**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sovrasov** [comment]
@Wovchena please have a look at python stuff. IS tiler is not fully finished (XAI is lacking). Also, tests are missing.

File: model_api/python/openvino/model_api/tilers/instance_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+             merged prediciton
+        """
+
+        if meta is None:
+            meta = {}
+        detection_result = super()._merge_results(results, shape, meta)
+
+        masks = []
+        for result in results:
+            if len(result["bboxes"]):
```
Comment (Wovchena) [inline model_api/python/openvino/model_api/tilers/instance_segmentation.py:115]: This should probably be
```suggestion
            if len(result["masks"]):
```

**Review by Wovchena — CHANGES_REQUESTED** [review]
All in all the design looks good

File: model_api/python/openvino/model_api/tilers/detection.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+        if np.prod(detections_array.shape):
+            detections_array, keep_idx = _multiclass_nms(
+                detections_array, max_num=self.max_pred_number
+            )
+            if meta is not None:
+                meta["keep_idx"] = keep_idx
+
+        merged_vector = (
+            np.average(feature_vectors, axis=0) if feature_vectors else np.ndarray(0)
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/detection.py:113]: Right now there is no reason, probably, during the original experiments, weighting of features was used.

File: model_api/python/openvino/model_api/tilers/detection.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+            detections = _detection2array(predictions.objects)
+            output_dict["saliency_map"] = predictions.saliency_map
+            output_dict["features"] = predictions.feature_vector
+        else:
+            detections = _detection2array(predictions)
+
+        offset_x, offset_y = meta["coord"][:2]
+        detections[:, 2:] += np.tile((offset_x, offset_y), 2)
+        output_dict["bboxes"] = detections
+        output_dict["tile_coords"] = meta["coord"]
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/detection.py:78]: There was a mistake in _predict_sync, meta should always be a dict

File: model_api/python/openvino/model_api/tilers/detection.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        for i, saliency_map in enumerate(saliency_maps[1:], 1):
+            for class_idx in range(num_classes):
+                if len(saliency_map.shape) == 4:
+                    saliency_map = saliency_map.squeeze(0)
+
+                cls_map = saliency_map[class_idx]
+
+                x_1, y_1, x_2, y_2 = tiles_coords[i]
+                y_1, x_1 = ((y_1, x_1) * ratio).astype(np.uint16)
+                y_2, x_2 = ((y_2, x_2) * ratio).astype(np.uint16)
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/detection.py]: Updated to use pure python

File: model_api/python/openvino/model_api/tilers/tiler.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+        tile_coords = self._tile(inputs)
+        tile_coords = self._filter_tiles(inputs, tile_coords)
+
+        if self.execution_mode == "sync":
+            return self._predict_sync(inputs, tile_coords)
+        return self._predict_async(inputs, tile_coords)
+
+    def _tile(self, image):
+        """Tiles an input image to either overlapping, non-overlapping or random patches.
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/tiler.py]: Updated

File: model_api/python/openvino/model_api/tilers/tiler.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+        By default, this method also adds the full image as the first tile to process.
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/tiler.py]: Updated

File: model_api/python/openvino/model_api/tilers/detection.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply model_api/python/openvino/model_api/tilers/detection.py]: Now you always pass `coord`. It makes sense to make it ordinary function arg.

File: model_api/python/openvino/model_api/tilers/tiler.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        Args:
+            predictions: model-dependent set of predicitons or one prediciton.
+            meta: a dict containing meta information required for postprocessing
+
+        Returns:
+            Postprocessed predictions
+        """
+
+    @abc.abstractmethod
+    def _merge_results(self, results, shapem, meta=None):
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/tiler.py:266]: Actually, it called in this way in IS tiler to save nms keep indices to meta. I'll try to split IS and detection implementations.

File: model_api/python/openvino/model_api/tilers/instance_segmentation.py (hunk 1 above)
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/instance_segmentation.py]: Fixed

File: model_api/python/openvino/model_api/tilers/detection.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/detection.py]: Changed

File: model_api/python/openvino/model_api/tilers/tiler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/tiler.py:263]: Changed, now IS tiler doesn't need that

File: model_api/python/openvino/model_api/tilers/detection.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+                if (map_h > y_2 - y_1 > 0) and (map_w > x_2 - x_1 > 0):
+                    cls_map = cv.resize(cls_map, (x_2 - x_1, y_2 - y_1))
+
+                map_h, map_w = y_2 - y_1, x_2 - x_1
+
+                for hi, wi in [(h_, w_) for h_ in range(map_h) for w_ in range(map_w)]:
+                    map_pixel = cls_map[hi, wi]
+                    merged_pixel = merged_map[class_idx][y_1 + hi, x_1 + wi]
+                    if merged_pixel != 0:
+                        merged_map[class_idx][y_1 + hi, x_1 + wi] = 0.5 * (map_pixel + merged_pixel)
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/detection.py]: Yes, that's a heuristic for merging. Max looks more natural, but during the initial experiments it produced more artifacts. I'll have a look at this one more time.

File: model_api/python/openvino/model_api/tilers/detection.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+             a dict with postprocessed predictions in 6-items format: (label id, score, bbox)
+        """
+
+        output_dict = {}
+        if hasattr(predictions, "objects"):
+            detections = _detection2array(predictions.objects)
+            output_dict["saliency_map"] = predictions.saliency_map
+            output_dict["features"] = predictions.feature_vector
+        else:
+            detections = _detection2array(predictions)
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/detection.py]: Updated

File: model_api/python/openvino/model_api/tilers/tiler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply model_api/python/openvino/model_api/tilers/tiler.py:263]: I can't find any call of `_merge_results()` with non default `meta`. Should `meta` be removed now as it is always `None`?

File: model_api/python/openvino/model_api/tilers/__init__.py
Comment (Wovchena) [inline model_api/python/openvino/model_api/tilers/__init__.py:1]: Should I merge this PR now?

**Review by Wovchena — APPROVED** [review]

File: model_api/python/openvino/model_api/tilers/__init__.py
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/__init__.py:1]: Yes, it's ready for merge. I'm working on cpp part in another branch

File: tests/python/accuracy/test_accuracy.py [hunk 10]
```diff
@@ -55,6 +67,8 @@ def test_image_models(data, dump, result, model_data):
         image = cv2.imread(str(image_path))
         if image is None:
             raise RuntimeError("Failed to read the image")
+        if "input_res" in model_data:
+            image = cv2.resize(image, eval(model_data["input_res"]))
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply tests/python/accuracy/test_accuracy.py:71]: The model was trained on that data and retrieves a very long list of detections, so the resulting json gets overwhelmed.

File: model_api/python/openvino/model_api/tilers/instance_segmentation.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+                masks.extend(result["masks"])
+
+        keep_idxs = []
+        if np.prod(detections_array.shape):
+            detections_array, keep_idxs = _multiclass_nms(
+                detections_array, max_num=self.max_pred_number
+            )
+
+        if masks:
+            masks = [masks[keep_idx] for keep_idx in keep_idxs]
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/tilers/instance_segmentation.py]: Removed that check

**Pull request merged**